arangod/IResearch/IResearchAqlAnalyzer.cpp

```diff
@@ -0,0 +1,516 @@
+    return _resolver;
+  }
+
+  virtual arangodb::velocypack::Options const& vpackOptions() const override {
+    return arangodb::velocypack::Options::Defaults;
+  }
+
+  /// @brief create a transaction::Context
+  virtual std::shared_ptr<arangodb::transaction::Context> newTrxContext() const override {
+    return std::shared_ptr<arangodb::transaction::Context>(
+        std::shared_ptr<arangodb::transaction::Context>(), &_transactionContext);
+  }
+
+  virtual arangodb::transaction::Methods& trxForOptimization() override {
+    return *_trx;
+  }
+
+  virtual bool killed() const override { return false; }
+
+  /// @brief whether or not a query is a modification query
+  virtual bool isModificationQuery() const noexcept override { return false; }
+
+  virtual bool isAsyncQuery() const noexcept override { return false; }
+
+  virtual void enterV8Context() override {
+    TRI_ASSERT(FALSE);
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+        TRI_ERROR_NOT_IMPLEMENTED,
+        "CalculationQueryContext entering V8 context is not implemented");
+  }
+
+ private:
+  arangodb::aql::QueryOptions _queryOptions;
+  arangodb::CollectionNameResolver _resolver;
+  mutable CalculationTransactionContext _transactionContext;
+  std::unique_ptr<arangodb::transaction::Methods> _trx;
+};
+
+frozen::set<irs::string_ref, 4> forbiddenFunctions{"TOKENS", "NGRAM_MATCH",
+                                                   "PHRASE", "ANALYZER"};
+
+arangodb::Result validateQuery(std::string const& queryStringRaw, TRI_vocbase_t& vocbase) {
+  try {
+    CalculationQueryContext queryContext(vocbase);
+    auto queryString = arangodb::aql::QueryString(queryStringRaw);
+    auto ast = queryContext.ast();
+    TRI_ASSERT(ast);
+    Parser parser(queryContext, *ast, queryString);
+    parser.parse();
+    ast->validateAndOptimize(queryContext.trxForOptimization());
+    AstNode* astRoot = const_cast<AstNode*>(ast->root());
+    TRI_ASSERT(astRoot);
+    // Forbid all V8 related stuff as it is not available on DBServers where analyzers run.
+    if (ast->willUseV8()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "V8 usage is forbidden for calculation analyzer"};
+    }
+
+    // no modification (as data access is forbidden) but to give more clear error message
+    if (ast->containsModificationNode()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "DML is forbidden for calculation analyzer"};
+    }
+
+    // no traversal (also data access is forbidden) but to give mo clear error message
+    if (ast->containsTraversal()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "Traversal usage is forbidden for calculation analyzer"};
+    }
+
+    std::string errorMessage;
+    // Forbid to use functions that reference analyzers -> problems on recovery as analyzers are not available for querying.
+    // Forbid all non-Dbserver runable functions as it is not available on DBServers where analyzers run.
+    arangodb::aql::Ast::traverseReadOnly(
```

Comment:
It is unclear to me why an "aql" analyzer so much AQL parts to be used here, e.g. LIMIT, FOR loops, SORT etc.
I don't think we should allow this, because it may lead to users doing very unexpected things with it.
Can't we strict the "aql" analyzer capabilties to AQL expressions, which are initiated by a RETURN statement, e.g. `RETURN @param + 1` or `RETURN SOUNDEX(REVERSE(@param))` or such?
That is IMHO the only way to effectively limit what users can do with the analyzer. Leaving it all open and then adding a few checks here works now, but in the future we may forget to add restrictions to the "aql" analyzer when we add new features.

Reply:
Ok, I`ll switch validation to 'whitelist' mode - all nodes except listed are forbidden.

Reply:
Should we allow all of this? I would favor restricting things to just `RETURN expression`, where `expression` would also be very limited. That would allow us to store the expression in an Aql `Expression` object, without the need to create execution plans etc.